tools/wake-format/emitter.cpp

```diff
@@ -39,123 +59,75 @@ wcl::rope Emitter::newline(ctx_t ctx) {
 wcl::rope Emitter::space(uint8_t count) {
   wcl::rope_builder builder;
   for (uint8_t i = 0; i < count; i++) {
-    builder.append("·");
+    builder.append(" ");
   }
   return std::move(builder).build();
 }
 
 wcl::optional<wcl::rope> Emitter::flat(ctx_t ctx, CSTElement node) {
-  return walk_node(ctx.flat(), node);
-}
-
-wcl::rope Emitter::try_flat(ctx_t ctx, CSTElement node) {
-  auto flat_node = flat(ctx, node);
-  if (flat_node && flat_node->column() <= max_column_width) {
-    return *flat_node;
+  // Try to get the flat version of node
+  auto flat_node = walk_node(ctx.flat(), node);
+  if (!flat_node) {
+    return {};
   }
-  auto full_node = walk_node(ctx, node);
-  assert(full_node);
-  return *full_node;
-}
 
-#define ASSERT_TOKEN(node, token) \
-  assert(!node.empty());          \
-  assert(node.id() == token);
-
-wcl::rope Emitter::layout(CST cst) {
-  ctx_t ctx;
-  return walk_top(ctx, cst.root()).concat(newline(ctx));
-}
-
-wcl::rope Emitter::walk_top(ctx_t ctx, CSTElement node) {
-  wcl::rope_builder builder;
-
-  for (CSTElement child = node.firstChildElement(); !child.empty(); child.nextSiblingElement()) {
-    if (child.isNode()) {
-      builder.append(try_flat(ctx, child));
-      builder.append(newline(ctx));
-    } else {
-      builder.append(walk_token(ctx, child));
-    }
+  if (flat_node->column() > max_column_width) {
+    return {};
   }
 
-  // probably shouldn't do this unconditionally
-  builder.undo();
-
-  return std::move(builder).build();
-}
-
-wcl::rope Emitter::walk_block(ctx_t ctx, CSTElement node) {
-  ASSERT_TOKEN(node, CST_BLOCK);
-  return walk_top(ctx.nest(), node);
+  return flat_node;
 }
 
-wcl::rope Emitter::walk_apply(ctx_t ctx, CSTElement node) {
-  ASSERT_TOKEN(node, CST_APP);
-
-  wcl::rope_builder builder;
-
-  CSTElement child = node.firstChildElement();
-  ASSERT_TOKEN(child, CST_ID);
-  {
-    auto node = walk_node(ctx, child);
-    assert(node);
-    builder.append(*node);
+wcl::optional<wcl::rope> Emitter::flat_or(ctx_t ctx, CSTElement node) {
```

Comment:
If you had `builder` ref here you could possible make this work for some context but you really need that + a function to call. Not sure you can make a flat_or function with this API but you can still write "flat or" code

```
auto flat_foo = walk_foo(ctx.flat(), foo_node);
if (flat_foo && builder.column() + flat_foo.size() <= max_column) {
  builder.append(*flat_foo);
} else {
  ctx_t sub_ctx = ctx.indent(); // Sometimes you might not want to indent here
  builder.append(newline(sub_ctx));
  builder.append(walk_foo(sub_ctx));
}
```

The things that change are the `walk_foo` and weather you want to indent or not. Two macros is probably sufficient, one for indent and one for not indenting.